src/main/scala/scorex/transaction/NgHistoryReader.scala

```diff
@@ -0,0 +1,76 @@
+package scorex.transaction
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import com.wavesplatform.state2._
+import scorex.block.Block.BlockId
+import scorex.block.{Block, MicroBlock}
+import scorex.transaction.History.BlockchainScore
+import scorex.utils.ScorexLogging
+
+class NgHistoryReader(ngState: () => Option[NgState], inner: History) extends History with NgHistory with DebugNgHistory with ScorexLogging with Instrumented {
+
+  override def synchronizationToken: ReentrantReadWriteLock = inner.synchronizationToken
+
+  override def height(): Int = read { implicit l =>
+    inner.height() + ngState().map(_ => 1).getOrElse(0)
+  }
+
+  override def blockBytes(height: Int): Option[Array[Byte]] = read { implicit l =>
+    inner.blockBytes(height).orElse(if (height == inner.height() + 1) ngState().map(_.bestLiquidBlock.bytes) else None)
+  }
+
+  override def scoreOf(blockId: BlockId): Option[BlockchainScore] = read { implicit l =>
+    inner.scoreOf(blockId)
+      .orElse(ngState() match {
+        case Some(ng) => Some(inner.score() + ng.base.blockScore)
+        case _ => None
```

Comment:
Hmm.. `ngState().map { ng => inner.score() + ng.base.blockScore }` ?

Reply:
should be `case Some(ng) if ng.contains(blockId) => Some(inner.score() + ng.base.blockScore)`